flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java

```diff
@@ -326,6 +366,29 @@ private void checkAndPropagateAsyncError() throws Exception {
 		}
 	}
 
+	/**
+	 * If the internal queue of the {@link KinesisProducer} gets too long,
+	 * flush some of the records until we are below the limit again.
+	 * We don't want to flush _all_ records at this point since that would
+	 * break record aggregation.
+	 */
+	private void enforceQueueLimit() {
```

Comment:
I wonder whether we could adjust the queue limit dynamically. 
you mentioned that `queue limit = (number of shards * queue size per shard) / record size`.
except record size, all others are relatively easy to set. For me, I don't really know the record size until the application starts. Also, what is the record size varies over time?
So how about add a queueLimit supplier function here to allow user to supply how the queueLimit is calculated dynamically?

Reply:
Btw, I am not requesting any changes, pr looks good to me for it defined purpose.

Just wonder how to config easily. Now I think about this a bit more. Will it be better if we expose `queue size` to the user instead  of `queue limit (number)`, thus, Inside of the FKP class, define an integer recordSize, and inside of the invoke function, do a moving average calculation of the recordSize with `serialized.remaining()` dynamically.

Reply:
A user-provided queue limit supplier function sounds like a good idea.
As you mentioned, this can come as a follow-up PR.

Reply:
The moving average calculation, that you described, could maybe just be a implementation of the limit supplier function.